src/app/reporting/ReportScheduler.h

```diff
@@ -78,29 +76,49 @@ class ReportScheduler : public ReadHandler::Observer
             // the scheduler in the ReadHandler
             Timestamp now = mTimerDelegate->GetCurrentMonotonicTimestamp();
             return (mReadHandler->IsGeneratingReports() &&
-                    ((now >= mMinTimestamp && mReadHandler->IsDirty()) || now >= mMaxTimestamp));
+                    (now >= mMinTimestamp && (mReadHandler->IsDirty() || now >= mMaxTimestamp || now >= mSyncTimestamp)));
         }
 
+        bool IsEngineRunScheduled() const { return mEnginRunScheduled; }
+        void SetEngineRunScheduled(bool aEnginRunScheduled) { mEnginRunScheduled = aEnginRunScheduled; }
+
         void SetIntervalTimeStamps(ReadHandler * aReadHandler)
         {
             uint16_t minInterval, maxInterval;
             aReadHandler->GetReportingIntervals(minInterval, maxInterval);
-            Timestamp now = mTimerDelegate->GetCurrentMonotonicTimestamp();
-            mMinTimestamp = now + System::Clock::Seconds16(minInterval);
-            mMaxTimestamp = now + System::Clock::Seconds16(maxInterval);
+            Timestamp now  = mTimerDelegate->GetCurrentMonotonicTimestamp();
+            mMinTimestamp  = now + System::Clock::Seconds16(minInterval);
+            mMaxTimestamp  = now + System::Clock::Seconds16(maxInterval);
+            mSyncTimestamp = mMaxTimestamp;
+        }
+
+        void RunCallback()
+        {
+            mScheduler->ReportTimerCallback();
+            SetEngineRunScheduled(true);
         }
 
-        void RunCallback() { mCallback(); }
+        void SetSyncTimestamp(System::Clock::Timestamp aSyncTimestamp)
+        {
+            // Prevents the sync timestamp being set to a value lower than the min timestamp
+            VerifyOrReturn(aSyncTimestamp >= mMinTimestamp);
+            mSyncTimestamp = aSyncTimestamp;
+        }
 
-        Timestamp GetMinTimestamp() const { return mMinTimestamp; }
-        Timestamp GetMaxTimestamp() const { return mMaxTimestamp; }
+        System::Clock::Timestamp GetMinTimestamp() const { return mMinTimestamp; }
+        System::Clock::Timestamp GetMaxTimestamp() const { return mMaxTimestamp; }
+        System::Clock::Timestamp GetSyncTimestamp() const { return mSyncTimestamp; }
 
     private:
         TimerDelegate * mTimerDelegate;
-        TimerCompleteCallback mCallback;
         ReadHandler * mReadHandler;
+        ReportScheduler * mScheduler;
         Timestamp mMinTimestamp;
         Timestamp mMaxTimestamp;
+        Timestamp mSyncTimestamp; // Timestamp at which the read handler will be allowed to emit a report so it can be synced with
+                                  // other handlers that have an earlier max timestamp
+        bool mEnginRunScheduled = false; // Flag to indicate if the engine run is already scheduled so the scheduler can ignore
```

Comment:
It really means if the ReadHandler will generate a report in an ongoing engine run. So if a run is scheduled, but a readHandler does not have this flag, its minimal timestamp can contribute in calculating the next timeout in the event the handler is currently dirty.

Reply:
Issue here for changes to name/behavior in future, will update comment to clarify in the meantime:
https://github.com/project-chip/connectedhomeip/issues/28052

Reply:
> It really means if the ReadHandler will generate a report in an ongoing engine run.

I am not sure how that's determined.  This can change mid-run as things become dirty, right?